fastapi-python-web-apis/main.py

```diff
@@ -0,0 +1,174 @@
+import random
+from typing import Annotated
+
+from pydantic import BaseModel, Field
+
+from fastapi import FastAPI, HTTPException, Query
+from fastapi.middleware.cors import CORSMiddleware
+
+tags_metadata = [
+    {
+        "name": "Random Playground",
+        "description": "Generate random numbers",
+    },
+    {
+        "name": "Random Items Management",
+        "description": "Create, shuffle, read, update and delete items",
+    },
+]
+
+app = FastAPI(
+    title="Randomizer API",
+    description="Shuffle lists, pick random items, generate random numbers.",
+    version="1.0.0",
+    openapi_tags=tags_metadata,
+)
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=["http://localhost:3000", "https://example.com"],
+    allow_credentials=True,
+    allow_methods=["GET", "POST", "PUT", "DELETE"],
+    allow_headers=["*"],
+)
+
+items_db = []
+
+
+class Item(BaseModel):
+    name: str = Field(
+        min_length=1, max_length=100, description="The item name"
+    )
+
+
+class ItemResponse(BaseModel):
+    message: str
+    item: str
+
+
+class ItemListResponse(BaseModel):
+    original_order: list[str]
+    randomized_order: list[str]
+    count: int
+
+
+class ItemUpdateResponse(BaseModel):
+    message: str
+    old_item: str
+    new_item: str
+
+
+class ItemDeleteResponse(BaseModel):
+    message: str
+    deleted_item: str
+    remaining_items_count: int
+
+
+@app.get("/", tags=["Random Playground"])
+async def home():
+    return {"message": "Welcome to the Randomizer API"}
+
+
+@app.get("/random/{max_value}")
+async def get_random_number(max_value: int, tags=["Random Playground"]):
+    return {"max": max_value, "random_number": random.randint(1, max_value)}
+
+
+@app.get("/random-between", tags=["Random Playground"])
+def get_random_number_between(
+    min_value: Annotated[
+        int | None,
+        Query(
+            title="Minimum Value",
+            description="The minimum random number",
+            ge=1,
+            le=1000,
+        ),
+    ] = 1,
+    max_value: Annotated[
+        int | None,
+        Query(
+            title="Maximum Value",
+            description="The maximum random number",
+            ge=1,
+            le=1000,
+        ),
+    ] = 99,
+):
+    if min_value > max_value:
+        raise HTTPException(
+            status_code=400, detail="min_value can't be greater than max_value"
+        )
+
+    return {
+        "min": min_value,
+        "max": max_value,
+        "random_number": random.randint(min_value, max_value),
+    }
+
+
+@app.post(
+    "/items", response_model=ItemResponse, tags=["Random Items Management"]
+)
+def add_item(item: Item):
```

Comment:
Should all of these be `async def`? Currently, only the first two are, same as you show in the tutorial. But you also mention that readers could update more to async, and that it's important to know which functions to asyncify and which ones not.

So, just wondering if that's intentional or leftover. I guess from my understanding, none of these endpoints benefit from being async (def also not `home()`) :P So just thinking that the code as a "best practice" should likely either commit to showing `async def` for all, or none, to avoid people wondering whether they've missed a subtle detail. Unless I _have_ missed one, of course! :)